ios/pcap/ipfinder.go

```diff
@@ -13,81 +16,172 @@ type NetworkInfo struct {
-func (n NetworkInfo) complete() bool {
-	return n.IPv6 != "" && n.Mac != "" && n.IPv4 != ""
+type PacketInfo struct {
+	Mac  string
+	IPv4 string
+	IPv6 string
+}
+
+func (n PacketInfo) complete() bool {
+	return (n.IPv6 != "" && n.Mac != "") || (n.IPv4 != "" && n.Mac != "")
 }
 
 // FindIp reads pcap packets until one is found that matches the given MAC
 // and contains an IP address. This won't work if the iOS device "automatic Wifi address" privacy
 // feature is enabled. The MAC needs to be static.
-func FindIp(device ios.DeviceEntry) (NetworkInfo, error) {
+func FindIPByMac(device ios.DeviceEntry) (NetworkInfo, error) {
 	mac, err := ios.GetWifiMac(device)
+	_ = mac
 	if err != nil {
 		return NetworkInfo{}, err
 	}
-	return findIp(device, mac)
+
+	// channel to receive PacketInfo
+	c := make(chan PacketInfo)
+	done := make(chan bool)
+
+	go func() {
+		if err := startCapture(device, c, done); err != nil {
+			log.Error(err)
+		}
+	}()
+
+	// captureDuration := time.Duration(10 * time.Second)
+	// startTime := time.Now()
+	for packet := range c {
+		if packet.Mac == mac {
+			done <- true
+			return NetworkInfo{Mac: packet.Mac, IPv4: packet.IPv4, IPv6: packet.IPv6}, nil
+		}
+	}
+
+	return NetworkInfo{}, nil
 }
 
-func findIp(device ios.DeviceEntry, mac string) (NetworkInfo, error) {
-	intf, err := ios.ConnectToService(device, "com.apple.pcapd")
+// FindIp reads pcap packets until one is found that matches the given MAC
+// and contains an IP address. This won't work if the iOS device "automatic Wifi address" privacy
+// feature is enabled. The MAC needs to be static.
+func FindIPByLazy(device ios.DeviceEntry, capture_duration time.Duration) (NetworkInfo, error) {
+	mac, err := ios.GetWifiMac(device)
+	_ = mac
 	if err != nil {
 		return NetworkInfo{}, err
 	}
+
+	// channel to receive PacketInfo
+	c := make(chan PacketInfo)
+	done := make(chan bool)
+
+	go func() {
+		if err := startCapture(device, c, done); err != nil {
+			log.Error(err)
+		}
+	}()
+
+	// captureDuration := time.Duration(10 * time.Second)
+	startTime := time.Now()
+	ipv6Hits := make(map[string]int)
+	ipv4Hits := make(map[string]int)
+	for packet := range c {
+		if time.Now().Sub(startTime) > capture_duration {
+			done <- true
+			break
+		}
+
+		if packet.IPv4 != "" {
+			ipv4Hits[packet.IPv4] += 1
+		}
+		if packet.IPv6 != "" {
+			ipv6Hits[packet.IPv6] += 1
+		}
+	}
+
+	highestIPv4Hits, highestIPv4Addr := 0, ""
+	for ipv4, hits := range ipv4Hits {
+		if hits > highestIPv4Hits {
+			highestIPv4Hits = hits
+			highestIPv4Addr = ipv4
+		}
+	}
+
+	highestIPv6Hits, highestIPv6Addr := 0, ""
+	for ipv6, hits := range ipv6Hits {
+		if hits > highestIPv6Hits {
+			highestIPv6Hits = hits
+			highestIPv6Addr = ipv6
+		}
+	}
+
+	return NetworkInfo{IPv4: highestIPv4Addr, IPv6: highestIPv6Addr}, nil
+}
+
+func startCapture(device ios.DeviceEntry, c chan PacketInfo, done chan bool) error {
+	intf, err := ios.ConnectToService(device, "com.apple.pcapd")
+	if err != nil {
+		return err
```

Comment:
I recently learned that for debugging things later, in Golang it is a good practice to wrap errors and indicate the function name of where they happened. So rather than just returning err here and in other places, it is better to do sth like: 
return fmt.Errorf("startCapture: failed connecting to com.apple.pcapd with err: %w, err)

Reply:
Should be resolved in latest commit.